File: blueprints/07-rag-basic/typescript/src/retrieval.ts
/**
 * Vector retrieval for the RAG Basic blueprint.
 *
 * Embeds a query using the same model used during ingestion, performs a
 * cosine-similarity search in ChromaDB, and returns hydrated Chunk objects
 * with similarity scores.
 */

import Anthropic from "@anthropic-ai/sdk";
import { ChromaClient, Collection } from "chromadb";

import type { Chunk } from "./ingestion.js";

export interface RetrieverConfig {
  collectionName?: string;
  chromaHost?: string;
  chromaPort?: number;
  anthropicApiKey?: string;
}

/**
 * Retrieves the most relevant document chunks for a given query.
 *
 * Uses Anthropic's voyage-3 embedding model for query embedding and
 * ChromaDB for approximate nearest-neighbor search.
 *
 * @example
 * ```typescript
 * const retriever = new VectorRetriever({ collectionName: "documents" });
 * const chunks = await retriever.retrieve("What is RAG?", 5);
 * for (const chunk of chunks) {
 *   console.log(`[${chunk.metadata.similarity}] ${chunk.source}`);
 *   console.log(chunk.content.slice(0, 200));
 * }
 * ```
 */
export class VectorRetriever {
  private readonly anthropic: Anthropic;
  private readonly chroma: ChromaClient;
  private readonly collectionName: string;
  private collection: Collection | null = null;

  constructor(config: RetrieverConfig = {}) {
    this.collectionName = config.collectionName ?? process.env.COLLECTION_NAME ?? "documents";

    const host = config.chromaHost ?? process.env.CHROMA_HOST ?? "localhost";
    const port = config.chromaPort ?? Number(process.env.CHROMA_PORT ?? "8000");

    this.anthropic = new Anthropic({
      apiKey: config.anthropicApiKey ?? process.env.ANTHROPIC_API_KEY,
    });
    this.chroma = new ChromaClient({ path: `http://${host}:${port}` });
  }

  /** Lazily initialize and cache the ChromaDB collection. */
  private async getCollection(): Promise<Collection> {
    if (!this.collection) {
      this.collection = await this.chroma.getOrCreateCollection({
        name: this.collectionName,
        metadata: { "hnsw:space": "cosine" },
      });
    }
    return this.collection;
  }

  /**
   * Embed the query and return the top-K most similar chunks.
   *
   * @param query - The user's question or search string.
   * @param topK - Number of chunks to return. More chunks provide richer
   *               context but increase token cost and may introduce noise.
   * @returns Array of Chunk objects sorted by descending similarity score.
   *          Each chunk's `metadata` contains a `similarity` key in [0, 1].
   * @throws Error if the query is empty or the collection has no documents.
   */
  async retrieve(query: string, topK: number = 5): Promise<Chunk[]> {
    if (!query.trim()) {
      throw new Error("Query must not be empty.");
    }

    const collection = await this.getCollection();
    const count = await collection.count();

    if (count === 0) {
      throw new Error(
        `Collection '${this.collectionName}' is empty. ` +
          "Run ingestion before querying."
      );
    }

    // Clamp topK to available documents
    const effectiveTopK = Math.min(topK, count);

    // Embed the query using the same model as ingestion
    const embedResponse = await this.anthropic.embeddings.create({
      model: "voyage-3",
      input: [query],
    });
    const queryEmbedding = embedResponse.data[0].embedding;

    // Perform vector similarity search in ChromaDB
    const results = await collection.query({
      queryEmbeddings: [queryEmbedding],
      nResults: effectiveTopK,
      include: ["documents", "metadatas", "distances"] as any,
    });

    // ChromaDB returns lists-of-lists (one per query); we have one query
    const ids = results.ids[0] ?? [];
    const documents = results.documents[0] ?? [];
    const metadatas = results.metadatas[0] ?? [];
    // In cosine space: distance = 1 - similarity
    const distances = (results.distances?.[0] ?? []) as number[];

    const chunks: Chunk[] = ids.map((id, i) => {
      const similarity = 1 - (distances[i] ?? 0);
      return {
        id,
        content: documents[i] ?? "",
        source: String(metadatas[i]?.source ?? "unknown"),
        metadata: {
          ...(metadatas[i] as Record<string, string | number | boolean>),
          similarity: Math.round(similarity * 10000) / 10000,
        },
      };
    });

    return chunks;
  }

  /**
   * Return basic statistics about the collection.
   */
  async getCollectionStats(): Promise<{ collectionName: string; totalChunks: number }> {
    const collection = await this.getCollection();
    const count = await collection.count();
    return {
      collectionName: this.collectionName,
      totalChunks: count,
    };
  }
}

File: legacy/blueprints/07-rag-basic/python/src/retrieval.py
"""
Vector retrieval for the RAG Basic blueprint.

Embeds a query using the same model used during ingestion, performs a
cosine-similarity search in ChromaDB, and returns hydrated Chunk objects
with similarity scores.
"""

from __future__ import annotations

import os
from typing import Any

import anthropic
import chromadb

from .ingestion import Chunk


class VectorRetriever:
    """
    Retrieves the most relevant document chunks for a given query.

    Uses Anthropic's voyage-3 embedding model for query embedding and
    ChromaDB for approximate nearest-neighbor search.

    Example usage::

        retriever = VectorRetriever(collection_name="documents")
        chunks = retriever.retrieve("What is retrieval-augmented generation?", top_k=5)
        for chunk in chunks:
            print(f"[{chunk.metadata['similarity']:.3f}] {chunk.source}")
            print(chunk.content[:200])
    """

    def __init__(
        self,
        collection_name: str = "documents",
        chroma_host: str | None = None,
        chroma_port: int | None = None,
        anthropic_api_key: str | None = None,
    ) -> None:
        """
        Args:
            collection_name: Name of the ChromaDB collection to search.
            chroma_host: ChromaDB server host. Defaults to CHROMA_HOST env var or 'localhost'.
            chroma_port: ChromaDB server port. Defaults to CHROMA_PORT env var or 8000.
            anthropic_api_key: Anthropic API key. Defaults to ANTHROPIC_API_KEY env var.
        """
        host = chroma_host or os.environ.get("CHROMA_HOST", "localhost")
        port = chroma_port or int(os.environ.get("CHROMA_PORT", "8000"))

        self._anthropic = anthropic.Anthropic(
            api_key=anthropic_api_key or os.environ["ANTHROPIC_API_KEY"]
        )
        self._chroma = chromadb.HttpClient(host=host, port=port)
        self._collection_name = collection_name
        self._collection = self._chroma.get_or_create_collection(
            name=collection_name,
            metadata={"hnsw:space": "cosine"},
        )

    def retrieve(self, query: str, top_k: int = 5) -> list[Chunk]:
        """
        Embed the query and return the top-K most similar chunks.

        Args:
            query: The user's question or search string.
            top_k: Number of chunks to return. More chunks provide more context
                   but increase token cost and may introduce noise.

        Returns:
            List of Chunk objects sorted by descending similarity score.
            Each chunk's ``metadata`` dict contains a ``similarity`` key with
            the cosine similarity score in [0, 1].

        Raises:
            ValueError: If the collection is empty.
            anthropic.APIError: If the embedding API call fails.
        """
        if not query.strip():
            raise ValueError("Query must not be empty.")

        # Check collection has documents
        count = self._collection.count()
        if count == 0:
            raise ValueError(
                f"Collection '{self._collection_name}' is empty. "
                "Run ingestion before querying."
            )

        # Clamp top_k to number of available documents
        effective_top_k = min(top_k, count)

        # Embed the query using the same model as ingestion
        response = self._anthropic.embeddings.create(
            model="voyage-3",
            input=[query],
        )
        query_embedding = response.data[0].embedding

        # Perform vector similarity search in ChromaDB
        results = self._collection.query(
            query_embeddings=[query_embedding],
            n_results=effective_top_k,
            include=["documents", "metadatas", "distances"],
        )

        # ChromaDB returns lists-of-lists (one per query); we only have one query
        ids: list[str] = results["ids"][0]
        documents: list[str] = results["documents"][0]
        metadatas: list[dict[str, Any]] = [dict(m) for m in (results["metadatas"][0] or [])]
        # ChromaDB cosine space returns distances where distance = 1 - similarity
        distances: list[float] = results["distances"][0]

        chunks: list[Chunk] = []
        for chunk_id, content, metadata, distance in zip(
            ids, documents, metadatas, distances, strict=False
        ):
            # Convert distance to similarity score for readability
            similarity = 1.0 - distance
            chunk = Chunk(
                id=chunk_id,
                content=content,
                source=metadata.get("source", "unknown"),
                metadata={
                    **metadata,
                    "similarity": round(similarity, 4),
                },
            )
            chunks.append(chunk)

        return chunks

    def get_collection_stats(self) -> dict:
        """Return basic statistics about the collection."""
        count = self._collection.count()
        return {
            "collection_name": self._collection_name,
            "total_chunks": count,
        }
